mobot/mobot_client/air_drop_session.py

```diff
@@ -0,0 +1,197 @@
+# Copyright (c) 2021 MobileCoin. All rights reserved.
+
+import random
+import full_service_cli as mc
+
+from decimal import Decimal
+from mobot_client.drop_session import BaseDropSession, SessionState
+from mobot_client.models import (
+    DropSession,
+    BonusCoin,
+)
+
+from mobot_client.chat_strings import ChatStrings
+
+
+class AirDropSession(BaseDropSession):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def handle_airdrop_payment(self, source, customer, amount_paid_mob, drop_session):
+        if not self.minimum_coin_available(drop_session.drop):
+            self.messenger.log_and_send_message(
+                customer,
+                source,
+                ChatStrings.AIRDROP_SOLD_OUT_REFUND.format(amount=amount_paid_mob.normalize())
+            )
+            self.send_mob_to_customer(customer, source, amount_paid_mob, True)
+            return
+
+        bonus_coin_objects_for_drop = BonusCoin.objects.filter(drop=drop_session.drop)
+        bonus_coins = []
+
+        for bonus_coin in bonus_coin_objects_for_drop:
+            number_claimed = DropSession.objects.filter(
+                drop=drop_session.drop, bonus_coin_claimed=bonus_coin
+            ).count()
+            number_remaining = bonus_coin.number_available - number_claimed
+            bonus_coins.extend([bonus_coin] * number_remaining)
+
+        if len(bonus_coins) <= 0:
+            self.messenger.log_and_send_message(
+                customer,
+                source,
+                ChatStrings.BONUS_SOLD_OUT_REFUND.format(amount=amount_paid_mob.normalize())
+            )
+            self.send_mob_to_customer(customer, source, amount_paid_mob, True)
+            return
+
+        initial_coin_amount_mob = mc.pmob2mob(
+            drop_session.drop.initial_coin_amount_pmob
+        )
+        random_index = random.randint(0, len(bonus_coins) - 1)
+        amount_in_mob = mc.pmob2mob(bonus_coins[random_index].amount_pmob)
+        amount_to_send_mob = (
+                amount_in_mob
+                + amount_paid_mob
+                + mc.pmob2mob(self.payments.get_minimum_fee_pmob())
+        )
+        self.send_mob_to_customer(customer, source, amount_to_send_mob, True)
+        drop_session.bonus_coin_claimed = bonus_coins[random_index]
+        drop_session.save()
```

Comment:
Can we remove the multiple `drop_session.save()` calls and just call it once at the end? Was there a reason to do all the intermediate commits to the DB vs. just using the object with changes in memory until the end of the method?

Reply:
@sugargoat Also fixed in my branch.